linux/desktop_webview_auth_plugin.cc

```diff
@@ -0,0 +1,175 @@
+#include "include/desktop_webview_auth/desktop_webview_auth_plugin.h"
+
+#include <flutter_linux/flutter_linux.h>
+#include <gtk/gtk.h>
+#include <glib.h>
+#include <cstring>
+
+#include <webkit2/webkit2.h>
+
+#define DESKTOP_WEBVIEW_AUTH_PLUGIN(obj)                                     \
+  (G_TYPE_CHECK_INSTANCE_CAST((obj), desktop_webview_auth_plugin_get_type(), \
+                              DesktopWebviewAuthPlugin))
+
+DesktopWebviewAuthPlugin *plugin;
+
+struct _DesktopWebviewAuthPlugin
+{
+  GObject parent_instance;
+  FlPluginRegistrar *registrar;
+  FlMethodChannel *method_channel;
+  gulong load_change_handler;
+  gchar *redirectUrl;
+  gchar *callbackUrl;
+  GtkWidget *popupWindow;
+  WebKitWebContext *context;
+};
+
+G_DEFINE_TYPE(DesktopWebviewAuthPlugin, desktop_webview_auth_plugin, g_object_get_type())
+
+static void changed(WebKitWebView *view, WebKitLoadEvent event, gpointer user_data)
+{
+  const gchar *uri = webkit_web_view_get_uri(view);
+  const gchar *redirectUrl = plugin->redirectUrl;
+
+  char str[(int)strlen(redirectUrl)];
+  const gchar *copy = strncpy(str, (char *)uri, (int)strlen(redirectUrl));
+
+  bool matching = strcmp(copy, redirectUrl) == 0;
+
+  if (event == WEBKIT_LOAD_FINISHED && matching)
+  {
+    webkit_web_context_clear_cache (plugin->context);
+
+    plugin->callbackUrl = (gchar *)webkit_web_view_get_uri(view);
+    g_autoptr(FlValue) result = fl_value_new_string(plugin->callbackUrl);
+
+    fl_method_channel_invoke_method(plugin->method_channel, "getCallbackUrl",
+                                    result, nullptr, nullptr, &plugin);
+
+    g_free(plugin->redirectUrl);
+    g_signal_handler_disconnect(G_OBJECT(view), plugin->load_change_handler);
+
+    gtk_window_close(GTK_WINDOW(plugin->popupWindow));
+  }
+}
+
+// Called if the user destroyed the window before sending the auth callback url.
+static void destroy(GtkWidget *widget, gpointer data)
+{
+  if (!plugin->callbackUrl)
+  {
+    fl_method_channel_invoke_method(plugin->method_channel, "getCallbackUrl",
+                                    nullptr, nullptr, nullptr, &plugin);
+  }
+}
+
+// Called when a method call is received from Flutter.
+static void desktop_webview_auth_plugin_handle_method_call(FlMethodCall *method_call)
```

Comment:
if you restore the first `DesktopWebviewAuthPlugin* self` argument, you have access to the plugin instance without using a global variable